File: slam/steps/buildOctree.mjs
// Порт BUILD_OCTREE (tasks.py:642-667) — переиспользует уже проверенный в
// этой сессии приём (bin/copc_convert_worker.ps1): pdal translate, с
// fallback на untwine при провале (живой случай на проде — PDAL
// STATUS_STACK_OVERFLOW на очень больших облаках, см. docs/PREPROCESSING.md).
// Финальные артефакты — пара las (сжатый LAZ) + copc (COPC-октодерево) —
// то же самое, что Python-версия кладёт в processed-bucket.
import { execFileSync } from "node:child_process";
import { existsSync, statSync, mkdirSync, renameSync, unlinkSync } from "node:fs";
import { runStep } from "../lib/stepContext.mjs";

const PDAL_EXE = "C:\\Users\\admin\\miniforge3\\envs\\geo\\Library\\bin\\pdal.exe";
const UNTWINE_EXE = "C:\\Users\\admin\\miniforge3\\envs\\geo\\Library\\bin\\untwine.exe";
const PROJ_DIR = "C:\\Users\\admin\\miniforge3\\envs\\geo\\Library\\share\\proj";
const PROJ_ENV = { ...process.env, PROJ_DATA: PROJ_DIR, PROJ_LIB: PROJ_DIR };

await runStep(async (ctx) => {
  const scanId = ctx.scan_id;
  const input = ctx.intermediate.georeference;
  mkdirSync(`${ctx.scan_dir}/final`, { recursive: true });

  const lasFinal = `${ctx.scan_dir}/final/pointcloud.laz`;
  const copcFinal = `${ctx.scan_dir}/final/pointcloud.copc.laz`;
  const srsArgs = ctx.crs_proj4 ? ["--writers.las.a_srs", ctx.crs_proj4] : [];

  // ---- final compressed LAZ (input LAS -> real LASzip compression) ----
  execFileSync(PDAL_EXE, ["translate", input, lasFinal, ...srsArgs], { env: PROJ_ENV, stdio: "pipe" });

  // ---- COPC octree: pdal translate, fallback to untwine (see header) ----
  const tmpCopc = copcFinal.replace(/\.copc\.laz$/, ".converting.copc.laz");
  let pdalError = null;
  try {
    execFileSync(PDAL_EXE, ["translate", input, tmpCopc, ...(ctx.crs_proj4 ? ["--writers.copc.a_srs", ctx.crs_proj4] : [])], { env: PROJ_ENV, stdio: "pipe" });
    if (!existsSync(tmpCopc)) throw new Error("pdal translate reported success but output is missing");
  } catch (e) {
    pdalError = e;
    if (existsSync(tmpCopc)) unlinkSync(tmpCopc);
  }
  if (pdalError) {
    try {
      execFileSync(UNTWINE_EXE, ["-i", input, "-o", tmpCopc], { env: PROJ_ENV, stdio: "pipe" });
    } catch (e2) {
      throw new Error(`pdal translate failed (${pdalError.message}); untwine fallback also failed: ${e2.message}`);
    }
    if (!existsSync(tmpCopc)) throw new Error("untwine fallback reported success but output is missing");
  }
  renameSync(tmpCopc, copcFinal);

  return {
    assets: [
      { type: "las", path: `${scanId}/final/pointcloud.laz`, size: statSync(lasFinal).size },
      { type: "copc", path: `${scanId}/final/pointcloud.copc.laz`, size: statSync(copcFinal).size },
    ],
  };
});

File: pyserver/apps/slam/pipeline/steps/build_octree.py
"""
Port of BUILD_OCTREE -- reuses the same pattern as
tours/management/commands/copc_convert_worker.py: pdal translate, falling
back to untwine on failure (real production case -- PDAL
STATUS_STACK_OVERFLOW on very large clouds). Final artifacts: a compressed
LAZ + a COPC octree, matching what the upstream Python pipeline puts in its
processed bucket.
"""

import os
import subprocess
from pathlib import Path

from django.conf import settings


def _pdal_proj_env():
    env = os.environ.copy()
    if settings.PDAL_PROJ_DATA_DIR:
        env["PROJ_DATA"] = settings.PDAL_PROJ_DATA_DIR
        env["PROJ_LIB"] = settings.PDAL_PROJ_DATA_DIR
    return env


def run(ctx: dict) -> dict:
    scan_id = ctx["scan_id"]
    input_path = ctx["intermediate"]["georeference"]
    final_dir = Path(ctx["scan_dir"]) / "final"
    final_dir.mkdir(parents=True, exist_ok=True)

    las_final = final_dir / "pointcloud.laz"
    copc_final = final_dir / "pointcloud.copc.laz"
    srs_args = ["--writers.las.a_srs", ctx["crs_proj4"]] if ctx.get("crs_proj4") else []
    env = _pdal_proj_env()

    # ---- final compressed LAZ (input LAS -> real LASzip compression) ----
    result = subprocess.run(
        [settings.PDAL_EXE, "translate", input_path, str(las_final), *srs_args],
        env=env, capture_output=True, text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(f"pdal translate (final LAZ) exited with code {result.returncode}: {result.stdout}{result.stderr}")

    # ---- COPC octree: pdal translate, fallback to untwine (see header) ----
    tmp_copc = str(copc_final)[: -len(".copc.laz")] + ".converting.copc.laz"
    pdal_error = None
    copc_srs_args = ["--writers.copc.a_srs", ctx["crs_proj4"]] if ctx.get("crs_proj4") else []
    result = subprocess.run(
        [settings.PDAL_EXE, "translate", input_path, tmp_copc, *copc_srs_args],
        env=env, capture_output=True, text=True,
    )
    if result.returncode != 0:
        pdal_error = f"pdal translate exited with code {result.returncode}: {result.stdout}{result.stderr}"
    elif not Path(tmp_copc).is_file():
        pdal_error = "pdal translate reported success but output is missing"
    if pdal_error:
        Path(tmp_copc).unlink(missing_ok=True)
        result = subprocess.run(
            [settings.UNTWINE_EXE, "-i", input_path, "-o", tmp_copc],
            env=env, capture_output=True, text=True,
        )
        if result.returncode != 0:
            raise RuntimeError(f"pdal translate failed ({pdal_error}); untwine fallback also exited with code {result.returncode}: {result.stdout}{result.stderr}")
        if not Path(tmp_copc).is_file():
            raise RuntimeError(f"pdal translate failed ({pdal_error}); untwine fallback reported success but output is missing")
    Path(tmp_copc).rename(copc_final)

    return {
        "assets": [
            {"type": "las", "path": f"{scan_id}/final/pointcloud.laz", "size": las_final.stat().st_size},
            {"type": "copc", "path": f"{scan_id}/final/pointcloud.copc.laz", "size": copc_final.stat().st_size},
        ],
    }

